Enable Interface tracing from earliest execution point

SimonWalton-HiFi opened the pull request:
This PR moves the --traceFile flag to early in main(), so as to allow tracing from the earliest point in the run.

`interface --traceFile FILE.json`
will write FILE.json to the user's Documents directory on exit.

`interface --traceFile FILE.json 12`
will exit and write file FILE.json to the user's Documents directory after twelve seconds.

Based on Clement's PR12700.

hifi-gustavo commented:
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/8681/app_12850_PR.apk

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/8681/HighFidelity-Beta-PR12850-ca5720c0158a8a3b261b32b53f7ccff9af908263.dmg

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/8681/HighFidelity-Beta-PR12850-ca5720c0158a8a3b261b32b53f7ccff9af908263.exe

SimonWalton-HiFi commented:
## Test Plan
`interface --traceFile trace.json`
should create a trace file ~/Documents/trace.json when the app quits.

`interface --traceFile trace.json 15`
should, in addition, automatically exit after 15 s.

The --traceScript option should work as before.

The trace files can be verified by dropping into chrome with 'chrome://tracing'.

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/8738/HighFidelity-Beta-PR12850-888fca2127531ae6901e4d1eeb1af54504f4d8a1.exe

hifi-gustavo commented:
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/8738/HighFidelity-Beta-Interface-PR12850-888fca2127531ae6901e4d1eeb1af54504f4d8a1.exe

samcake commented on libraries/task/src/task/Task.h:
```diff
@@ -326,6 +330,7 @@ class Task : public Job<JC> {
         }
 
         void applyConfiguration() override {
+           Duration profileRange(trace_render(), ("configure::" + JobConcept::getName()).c_str());
```
simple space

SimonWalton-HiFi replied:
Fixed.

samcake commented:
WE need a basic test plan for this

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/8743/HighFidelity-Beta-PR12850-5c141aa5e99a8d9ae50c112fbef31c4a058f6d2d.exe

hifi-gustavo commented:
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/8743/HighFidelity-Beta-Interface-PR12850-5c141aa5e99a8d9ae50c112fbef31c4a058f6d2d.exe

hifi-gustavo commented:
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/8745/app_12850_PR_a584d67acde3d39fd8cda702415529e0f2c2bfb9.apk

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/8745/HighFidelity-Beta-PR12850-a584d67acde3d39fd8cda702415529e0f2c2bfb9.dmg

hifi-gustavo commented:
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/8745/HighFidelity-Beta-Interface-PR12850-a584d67acde3d39fd8cda702415529e0f2c2bfb9.dmg

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/8745/HighFidelity-Beta-PR12850-a584d67acde3d39fd8cda702415529e0f2c2bfb9.exe

hifi-gustavo commented:
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/8745/HighFidelity-Beta-Interface-PR12850-a584d67acde3d39fd8cda702415529e0f2c2bfb9.exe

conklin94122 commented:
@SimonWalton-HiFi @Atlante45 The test plan needs to be clearer, and if --disableWatchdog needs to be tested those tests need to be enumerated and clarified.

Removing QA Ready until the test cases are reworked. Thanks.

Atlante45 commented on interface/src/Application.cpp:
```diff
@@ -958,21 +962,18 @@ Application::Application(int& argc, char** argv, QElapsedTimer& startupTimer, bo
-                setProperty(hifi::properties::TRACING, traceFilePath);
-                DependencyManager::get<tracing::Tracer>()->startTracing();
+            } else if (args.at(i) == QStringLiteral("--disableWatchdog")) {
+                watchdogCrashWhenTriggered = false;
```
There is already a `DISABLE_WATCHDOG` that exists. We should enable setting it from the command line.
I think this should be in the main with most of the other command line args.

SimonWalton-HiFi replied:
The DISABLE_WATCHDOG is a preprocessor flag.

Atlante45 replied:
No, it's just a global variable.

Atlante45 commented on interface/src/main.cpp:
```diff
@@ -40,6 +42,26 @@ extern "C" {
+            }
+            break;
+        }
+    }
```
This should be done in the `QCommandLineParser` bellow.
If we still want to enable the tracer that early, the the parsing should be moved up.

SimonWalton-HiFi commented:
I've changed the disableWatchdog flag to use the existing DISABLE_WATCHDOG variable, as requested. The effect is slightly different in that my earlier approach still ran a watchdog and gave warning, it just didn't abort on timeout. I've also changed to use the cmdOptionExists() routine like the other flags.

hifi-gustavo commented:
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/8771/app_12850_PR_76468f7f261184a96ee5967c973b22b3dad75d13.apk

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/8771/HighFidelity-Beta-PR12850-76468f7f261184a96ee5967c973b22b3dad75d13.dmg

hifi-gustavo commented:
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/8771/HighFidelity-Beta-Interface-PR12850-76468f7f261184a96ee5967c973b22b3dad75d13.dmg

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/8771/HighFidelity-Beta-PR12850-76468f7f261184a96ee5967c973b22b3dad75d13.exe

hifi-gustavo commented:
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/8771/HighFidelity-Beta-Interface-PR12850-76468f7f261184a96ee5967c973b22b3dad75d13.exe

conklin94122 commented:
@SimonWalton-HiFi I removed QA Ready for now. Once you have Dev Approval please put QA Ready back and the testers will do their thing. Thanks.

Qliemillar commented:
Currently being tested
